packages/next-upgrade/index.ts

```diff
@@ -0,0 +1,368 @@
+#!/usr/bin/env node
+
+import prompts from 'prompts'
+import fs from 'fs'
+import { execSync } from 'child_process'
+import path from 'path'
+import { compareVersions } from 'compare-versions'
+import chalk from 'chalk'
+import which from 'which'
+import { createRequire } from 'node:module'
+
+type StandardVersionSpecifier = 'canary' | 'rc' | 'latest'
+type CustomVersionSpecifier = string
+type VersionSpecifier = StandardVersionSpecifier | CustomVersionSpecifier
+type PackageManager = 'pnpm' | 'npm' | 'yarn' | 'bun'
+
+interface Response {
+  version: StandardVersionSpecifier
+}
+
+async function run(): Promise<void> {
+  const appPackageJsonPath = path.resolve(process.cwd(), 'package.json')
+  let appPackageJson = JSON.parse(fs.readFileSync(appPackageJsonPath, 'utf8'))
+
+  await detectWorkspace(appPackageJson)
+
+  let targetNextPackageJson
+  let targetVersionSpecifier: VersionSpecifier = ''
+
+  const shortcutVersion = process.argv[2]?.replace('@', '')
+  if (shortcutVersion) {
+    const res = await fetch(
+      `https://registry.npmjs.org/next/${shortcutVersion}`
+    )
+    if (res.status === 200) {
+      targetNextPackageJson = await res.json()
+      targetVersionSpecifier = targetNextPackageJson.version
+    } else {
+      console.error(
+        `${chalk.yellow('Next.js ' + shortcutVersion)} does not exist. Check available versions at ${chalk.underline('https://www.npmjs.com/package/next?activeTab=versions')}, or choose one from below\n`
+      )
+    }
+  }
+
+  if (!targetNextPackageJson) {
+    let nextPackageJson: { [key: string]: any } = {}
+    try {
+      const resCanary = await fetch(`https://registry.npmjs.org/next/canary`)
```

Comment:
Should we consider user's registry?
Possible case of using private registry or yarn has it's own (https://registry.yarnpkg.com).
x-ref: https://github.com/vercel/next.js/blob/canary/packages/next/src/lib/helpers/get-registry.ts